arangod/Replication2/StateMachines/Document/DocumentStateStrategy.cpp

```diff
@@ -133,4 +140,142 @@ auto DocumentStateShardHandler::createLocalShard(
   }
 
   return ResultT<std::string>::success(std::move(shardId));
-}
\ No newline at end of file
+}
+
+DocumentStateTransaction::DocumentStateTransaction(
+    std::unique_ptr<transaction::Methods> methods)
+    : _methods(std::move(methods)) {}
+
+auto DocumentStateTransaction::apply(DocumentLogEntry const& entry)
+    -> futures::Future<Result> {
+  // TODO revisit checkUniqueConstraintsInPreflight and waitForSync
+  auto opOptions = OperationOptions();
+  opOptions.silent = true;
+  opOptions.ignoreRevs = true;
+  opOptions.isRestore = true;
+  // opOptions.checkUniqueConstraintsInPreflight = true;
+  opOptions.validate = false;
+  opOptions.waitForSync = false;
+  opOptions.indexOperationMode = IndexOperationMode::internal;
+
+  auto fut =
+      futures::Future<OperationResult>{std::in_place, Result{}, opOptions};
+  switch (entry.operation) {
+    case kInsert:
+      fut = _methods->insertAsync(entry.shardId, entry.data.slice(), opOptions);
+      break;
+    case kUpdate:
+      fut = _methods->updateAsync(entry.shardId, entry.data.slice(), opOptions);
+      break;
+    case kReplace:
+      fut =
+          _methods->replaceAsync(entry.shardId, entry.data.slice(), opOptions);
+      break;
+    case kRemove:
+      fut = _methods->removeAsync(entry.shardId, entry.data.slice(), opOptions);
+      break;
+    case kTruncate:
+      // TODO Think about correctness and efficiency.
+      fut = _methods->truncateAsync(entry.shardId, opOptions);
+      break;
+    default:
+      return Result{
+          TRI_ERROR_TRANSACTION_INTERNAL,
+          fmt::format("Transaction of type {} with ID {} could not be applied",
+                      to_string(entry.operation), entry.tid.id())};
+  }
+
+  return std::move(fut).thenValue(
+      [self = shared_from_this()](OperationResult&& opRes) {
+        return opRes.result;
+      });
```

Comment:
In FollowerState we have the assertion that the Future isReady anyway, so this could be simplified to:
```
TRI_ASSERT(fut.isReady());
return std::move(fut).get().result;
```

Reply:
Will be addressed in https://github.com/arangodb/arangodb/pull/16650